src/main/java/org/truffleruby/parser/BodyTranslator.java

```diff
@@ -1704,7 +1705,13 @@ public RubyNode visitGlobalAsgnNode(GlobalAsgnParseNode node) {
         if (THREAD_LOCAL_GLOBAL_VARIABLES.contains(name)) {
             final GetThreadLocalsObjectNode getThreadLocalsObjectNode = GetThreadLocalsObjectNodeGen.create();
             getThreadLocalsObjectNode.unsafeSetSourceSection(sourceSection);
-            return addNewlineIfNeeded(node, withSourceSection(sourceSection, new WriteInstanceVariableNode(name, getThreadLocalsObjectNode, rhs)));
+            RubyNode writeNode;
+            if("$SAFE".equals(name)){
+                writeNode = new WriteSafeLevelNode(rhs);
+            } else {
+                writeNode = new WriteInstanceVariableNode(name, getThreadLocalsObjectNode, rhs);
+            }
```

Comment:
In the `switch` above, you should have `case "$SAFE": rhs = new CheckSafeLevelNode(rhs)`.
`CheckSafeLevelNode` can then just call `Rubinius::Type.set_safe_level` (or better `check_safe_level`).
And that would only do the check, not the write.
So then the logic here can remain the same.

Reply:
Fixed.